src/Guard.php

```diff
@@ -289,10 +289,22 @@ protected function getLastKeyPair(): ?array
             return null;
         }
 
-        end($this->storage);
-        $name = key($this->storage);
-        $value = $this->storage[$name];
-        reset($this->storage);
+        $name = null;
+        $value = null;
+        if (is_array($this->storage)) {
+            end($this->storage);
+            $name = key($this->storage);
+            $value = $this->storage[$name];
+            reset($this->storage);
+        } elseif ($this->storage instanceof Iterator) {
+            $this->storage->rewind();
```

Comment:
I don't know if this optimisation is worth it, but if `$this->storage` implements `Countable` and `SeekableIterator`, we could do:

```php
$this->storage->seek(count($this->storage) - 1); 
$name = $this->storage->key();
$value = $this->storage->current();
```

and avoid the loop. I suppose it depends how many items are likely to be in the storage as to whether this is worth it though.

Reply:
I think then there is a conflict with Guard.php:131. There is the Iterator interface the minimum dependency and not the SeekableIterator.

Reply:
Fair enough.